src/compiler/emitter.ts

```diff
@@ -4866,20 +4867,27 @@ export function createPrinter(printerOptions: PrinterOptions = {}, handlers: Pri
     }
 
     function emitEmbeddedStatement(parent: Node, node: Statement) {
-        if (isBlock(node) || getEmitFlags(parent) & EmitFlags.SingleLine) {
+        const lines = getLeadingLineTerminatorCount(parent, node, ListFormat.None);
+        if (isBlock(node) || getEmitFlags(parent) & EmitFlags.SingleLine || !lines) {
             writeSpace();
             emit(node);
         }
         else {
-            writeLine();
-            increaseIndent();
-            if (isEmptyStatement(node)) {
-                pipelineEmit(EmitHint.EmbeddedStatement, node);
+            if (lines) {
+                writeLine(lines);
+                increaseIndent();
+                if (isEmptyStatement(node)) {
+                    pipelineEmit(EmitHint.EmbeddedStatement, node);
+                }
+                else {
+                    emit(node);
+                }
+                decreaseIndent();
             }
             else {
+                writeSpace();
                 emit(node);

```

Comment:
Instead of changing only the behavior for `if` only, I found that `if, else, do, while, for, forin, forof, with` all calls `emitEmbeddedStatement` and thus should all so have the behavior of preserving newlines/spaces. Thus, changing the general behavior here.